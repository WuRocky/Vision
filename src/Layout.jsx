import { Outlet, Link, useLocation } from "react-router-dom";
import React, { useState, useEffect, useRef } from "react";
import Footer from "./components/Footer";

import {
    getData,
    getTopicsData,
    getPopularData,
    getMyData,
    getPopularAuthor,
} from "./hooks/useFireStore";

import { getUser } from "./hooks/useFireAuthentication";
import Nav from "./components/Nav";

const AppContext = React.createContext();

const Layout = () => {
    // 取得前兩邊文章
    const [firebaseTwoDoc, setFirebaseTwoDoc] = useState([]);

    // 熱門文章
    const [popularArticles, setPopularArticles] = useState([]);

    // 得到熱門作者
    const [popularAuthor, setPopularAuthor] = useState([]);

    // 使用者文章
    const [myArticles, setMyArticles] = useState([]);

    // 登入者資訊
    const [user, setUser] = useState(null);

    // 文章分類
    const [topics, setTopics] = useState([]);

    const [isLoading, setIsLoading] = useState(true);

    const location = useLocation();
    const urlSearchParams = new URLSearchParams(location.search);
    const currentTopics = urlSearchParams.get("topics");
    const articleRef = useRef();

    const [lastArticleRef, setLastArticleRef] = useState(articleRef.current);
    useEffect(() => {
        // 取得前兩邊文章
        getData(setFirebaseTwoDoc, currentTopics, setLastArticleRef);

        // 登入者資訊
        getUser(setUser);

        // 得到熱門作者
        getPopularAuthor(setPopularAuthor);

        // 熱門文章
        getPopularData(setPopularArticles);

        // 使用者文章
        getMyData(setMyArticles);

        // 文章分類
        getTopicsData().then((data) => {
            setTopics(data);
            setIsLoading(false);
        });
    }, [currentTopics]);
    if (isLoading) return <div>Loading...</div>;

    return (
        <AppContext.Provider
            value={{
                firebaseTwoDoc,
                setFirebaseTwoDoc,
                lastArticleRef,
                setLastArticleRef,
                user,
                setUser,
                topics,
                popularArticles,
                setPopularArticles,
                myArticles,
                popularAuthor,
            }}
        >
            <div>
                <Nav />
                <Outlet />
                <Footer />
            </div>
        </AppContext.Provider>
    );
};
export { AppContext };
export default Layout;
