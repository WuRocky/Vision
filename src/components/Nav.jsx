import React, { useContext } from "react";
import { Link } from "react-router-dom";
import Notes from "../img/notes2.png";
import { AppContext } from "../Layout";
import { userSignOut } from "../hooks/useFireAuthentication";
const Nav = () => {
    const { user } = useContext(AppContext);
    const logout = () => {
        userSignOut();
        window.location.reload();
    };
    return (
        <div className="navContainer">
            <div>
                <Link to="/">
                    <img src={Notes} alt="Vision" />
                    <p>Vision</p>
                </Link>
                <nav>
                    <ul>
                        {user ? (
                            <>
                                <li>
                                    <Link to="/write">Write</Link>
                                </li>
                                <li>
                                    <Link to="/member">Member</Link>
                                </li>
                                <li>
                                    <Link onClick={logout}>Sign out</Link>
                                </li>
                            </>
                        ) : (
                            <>
                                {/* <li>
                                    <Link to="/write">Write</Link>
                                </li> */}
                                <li>
                                    <Link className="signIn" to="/signIn">
                                        Register / Sign In
                                    </Link>
                                </li>
                            </>
                        )}
                    </ul>
                </nav>
            </div>
        </div>
    );
};

export default Nav;
