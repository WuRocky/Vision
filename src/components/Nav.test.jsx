import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { BrowserRouter as Router } from "react-router-dom";
import { render, screen } from "@testing-library/react";
import { AppContext } from "../Layout";
import Nav from "./Nav";
import renderer from "react-test-renderer"; // Add this import

const renderWithContext = (user) => {
    return render(
        <AppContext.Provider value={{ user }}>
            <Router>
                <Nav />
            </Router>
        </AppContext.Provider>
    );
};

describe("Nav component", () => {
    // Add the snapshot test for no user
    test("renders Nav component as a snapshot without user", () => {
        const tree = renderer
            .create(
                <AppContext.Provider value={{ user: null }}>
                    <Router>
                        <Nav />
                    </Router>
                </AppContext.Provider>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    // Add the snapshot test for logged-in user
    test("renders Nav component with logged-in user as a snapshot", () => {
        const user = { uid: "123", displayName: "John Doe" };
        const tree = renderer
            .create(
                <AppContext.Provider value={{ user }}>
                    <Router>
                        <Nav />
                    </Router>
                </AppContext.Provider>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    test("shows Register / Sign In link when user is not logged in", () => {
        renderWithContext(null);
        const signInLink = screen.getByText(/Register \/ Sign In/i);
        expect(signInLink).toBeInTheDocument();
    });

    test("shows Write, Member, and Sign out links when user is logged in", () => {
        const user = { uid: "123", displayName: "John Doe" };
        renderWithContext(user);
        const writeLink = screen.getByText(/Write/i);
        const memberLink = screen.getByText(/Member/i);
        const signOutLink = screen.getByText(/Sign out/i);

        expect(writeLink).toBeInTheDocument();
        expect(memberLink).toBeInTheDocument();
        expect(signOutLink).toBeInTheDocument();
    });
});
