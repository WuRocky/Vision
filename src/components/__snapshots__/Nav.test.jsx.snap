// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav component renders Nav component as a snapshot without user 1`] = `
<div
  className="navContainer"
>
  <div>
    <a
      href="/"
      onClick={[Function]}
    >
      <img
        alt="Vision"
        src=""
      />
      <p>
        Vision
      </p>
    </a>
    <nav>
      <ul>
        <li>
          <a
            className="signIn"
            href="/signIn"
            onClick={[Function]}
          >
            Register / Sign In
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;

exports[`Nav component renders Nav component with logged-in user as a snapshot 1`] = `
<div
  className="navContainer"
>
  <div>
    <a
      href="/"
      onClick={[Function]}
    >
      <img
        alt="Vision"
        src=""
      />
      <p>
        Vision
      </p>
    </a>
    <nav>
      <ul>
        <li>
          <a
            href="/write"
            onClick={[Function]}
          >
            Write
          </a>
        </li>
        <li>
          <a
            href="/member"
            onClick={[Function]}
          >
            Member
          </a>
        </li>
        <li>
          <a
            href="/"
            onClick={[Function]}
          >
            Sign out
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;
